### PR TITLE
chore(email): Switch to \Zend\Mail library instead of raw mail() function

### LINKED ISSUE #5918: Adopt third-party email library
Our email management is scary and hacky. There are plenty of good, object-oriented solutions out there that come with attachment support and such.

* [Symfony/Swiftmailer](http://symfony.com/doc/current/cookbook/email/email.html) (MIT)
* [Zend](https://github.com/zendframework/zf2/tree/master/library/Zend/Mail) (3-Clause BSD)
* [PHPMailer](https://github.com/PHPMailer/PHPMailer) (LGPL 2.1)

Right now I'm leaning Zend because it also has support for *receiving* emails, which would be amazing.

### PR DESCRIPTION
**Pull request by mrclay**
(replaces #8315)

This lays a good foundation for improving Elgg's flexibility and security in sending emails. Some of the features we open up because of this change:
- HTML emails
- Attachments
- Improved security (automatic header escaping, etc.)
- Third-party SMTP services (e.g. Mandrill)
- Receiving emails
- In-memory transport for easier testing/development

We aren't introducing our own interfaces because `Zend\Mail` is plenty well supported and the API is agreeable to us. Writing a wrapper would just be unnecessary overhead.

BREAKING CHANGE:
We are switching to `Zend\Mail` for sending emails in Elgg 2.0. It's likely that there are some edge cases that the library handles differently than Elgg 1.x used to. Take care to test your email notifications carefully when upgrading to 2.0.

Fixes #5918

### CODE REVIEW

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 1]
```diff
@@ -154,6 +156,9 @@ public function __construct(\Elgg\Config $config) {
 		$this->setFactory('logger', function(ServiceProvider $c) {
 			return $this->resolveLoggerDependencies('logger');
 		});
+		
+		// TODO(evan): Support configurable transports...
+		$this->setClassName('mailer', 'Zend\Mail\Transport\Sendmail');
```
Comment (ewinslow) [inline engine/classes/Elgg/Di/ServiceProvider.php]: Use `::class`

File: engine/lib/notification.php [hunk 2]
```diff
@@ -622,63 +625,40 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
...
-	}
+	$subject = trim($subject);
 
-	// Format message
-	$body = html_entity_decode($body, ENT_QUOTES, 'UTF-8'); // Decode any html entities
-	$body = elgg_strip_tags($body); // Strip tags from message
-	$body = preg_replace("/(\r\n|\r)/", "\n", $body); // Convert to unix line endings in body
-	$body = preg_replace("/^From/", ">From", $body); // Change lines starting with From to >From
+	$body = elgg_strip_tags($result['body']);
+	$body = html_entity_decode($body, ENT_QUOTES, 'UTF-8');
```
Comment (ewinslow) [inline engine/lib/notification.php]: This changes the order of strip tags and decode. Is that on purpose?

**Comment by ewinslow** [comment]
LGTM

File: engine/lib/notification.php (hunk 2 above)
Comment (mrclay) [inline engine/lib/notification.php]: That's the fix for https://github.com/Elgg/Elgg/issues/7480. I guess I can wait on that and submit another pr

Comment (ewinslow) [inline engine/lib/notification.php]: If that but is meant to be fixed then it least needs to be mentioned in the commit

Comment (ewinslow) [inline engine/lib/notification.php]: Which it is... Going to pull this in.
